cpp/src/arrow/engine/substrait/options.h

```diff
@@ -83,6 +84,14 @@ class ARROW_ENGINE_EXPORT ExtensionProvider {
 
 ARROW_ENGINE_EXPORT std::shared_ptr<ExtensionProvider> default_extension_provider();
 
+struct ARROW_ENGINE_EXPORT NamedTapNodeOptions : public compute::ExecNodeOptions {
+  NamedTapNodeOptions(const std::string& name, std::shared_ptr<Schema> schema)
+      : name(name), schema(std::move(schema)) {}
+
+  std::string name;
+  std::shared_ptr<Schema> schema;
+};
```

Comment:
I'm not sure what the plan is here.  I thought originally the goal was to mimic the named table.  In that case I'd expect to see something like `NamedTableProvider` (e.g. `NamedTapProvider`) that translates names into actual nodes (e.g. a `TeeNode`).

In that case I wouldn't expect there to be any "named tap" equivalent in Acero.  I'm not sure what `NamedTapNode` and `AddPassFactory` are providing in this PR?

Reply:
If you return exec node options (similar to named tap provider) then you could bypass the need to encode properties into the name (you wouldn't even really need `kind`):

```
# in python
def tap_provider(name):
  if name == 'one':
    return TeeNodeOptions('/tmp/dataset_one')
  elif name == 'two':
    return TeeNodeOptions('/tmp/dataset_two')
  else:
    raise Error(...)
```

or you could move the name encoding / decoding into python

```
# in python
def tap_provider(name)
  path = get_path_from_name(name)
  return TeeNodeOptions(path)
```

or you could still use the kind mapping

```
# in python
def tap_provider(name)
  kind = get_kind_from_name(name)
  if kind == 'tee':
   return TeeNodeOptions(path)...
```

However, this feature is still pretty experimental so I don't mind sticking with kind<->factory name mapping for now if that is what you would prefer.

Reply:
In my mind we should do the following:

(1) Add "NamedTapRel" as a purely abstract relation in Acero. The Acero code itself doesn't care or know how to turn a NamedTapRel message into a Declaration.

(2) Add a C++ extension API that allows downstream users / implementer of the NamedTapRel to tell Acero "hey, if you see a NamedTapRel message, call into the custom substrait extension function that I provide here and I will return a declaration to u" (basically this: https://github.com/apache/arrow/issues/33850)

This way I believe we don't need most the stuff here (including changes to pass conv_opts, named_tap_mapper, etc)

Reply:
@westonpace @rtpsw WDYT?

Reply:
Maybe the API for (2) can also be "Hey, if you see an certain type of message (is there a way to detect a message is an extension message)?) pass it to the custom extension function that I provide and I will return a declaration to u"

Reply:
I think with this approach, we can basically support any ExtensionSingleRel, ExtensionMultiRel and ExtensionLeafRel in the future with a single extension point (the custom substrait provider)

Reply:
Taking it a step further, we could just add `NamedTapRel` without any callback code, because one would be able (with #33850 and without this PR) register a custom extension provider that would handle it in `MakeRel` (and forward what it doesn't handle to the original/default extension provider). One step more, we could add nothing and just let the extension interpret an `Any` message. This feels like stepping in the wrong direction, which aggressively prioritizes generality. I feel we should have some structure in Acero for this, at least so that structures would get reused instead of reinvented.

Reply:
IMO I don't think this is over generalized - it matches the level of generalization in substrait extension protobuf - Leaf, single rel and multiple rel. As something experimental at this stage I think we do benefit from generalization since we make very minimal changes to Acero code itself and more free to experiment / iterator faster on our side.

Reply:
There's also the option of having both - a user would use the specific structures when they fit and the more general one when they don't. If we go with this approach, this PR will be for a specific structure. I'd also like to hear @westonpace's opinion.

Reply:
IFWIW,  agree a specific structure can be useful however in this case I don't feel it is much more useful than a general one in this case. For example, the specific structure for the NamedTapNode make function returns a NamedTapNode option, and doesn't not allow the user to return a declaration with a custom node option (e.g, in our case, a WriteSmoothNodeOption), so we need to handle that in our code after calling this NamedTapNode make function here (or totally by pass NamedTapNode function here). This an be done but I feel that is unnecessary indirection.

Reply:
I agree with @rtpsw that going too generic (protobuf.Any in and Declaration out) is not a good idea.  This is the exact API for the `ExtensionProvider` and so we already have this.  You could use this for `WriteSmoothNodeOption` and it would end up looking just like the extension that is in place for AsofJoinNode.  If that works for you then there is nothing we need to do :)

So if we want to go less generic then we should look at what Acero provides at each step to simplify things:

### Generic Node

The next step down in generality would be to remove protobuf from the picture using Substrait/Arrow literals:

```
message GenericOptions {
  message Property {
    string key = 0;
    Literal value = 1;
  }
  repeated Property properties = 1;
}

message GenericNode {
  string kind = 0;
  GenericOptions options = 1;
}
```

The API in Arrow would be:

```
void RegisterGenericExtensionHandler(string kind, function<Declaration(unordered_map<string, Scalar>)> handler)
```

In other words, given properties as a map of arrow scalars, create a declaration

**Pros:**
  * No need for extension author to write proto file
  * Can be used for any extension type
**Cons:**
  * Need to create an entire ExecNode for your custom behavior

### Generic Mapper

An even more focused approach could be to remove the need to create an exec node at all.  This would be focused on the 1 input in / 1 input out / no accumulation case (e.g. what MapNode handles).  From what I know of your write need this would fit this case.

The Substrait would be pretty much identical to the above.

```
message MapNode {
  string kind = 0;
  GenericOptions options = 1;
}
```

However, the Arrow API would be:

```
using Mapper = function<ExecBatch(ExecBatch)>;
void RegisterGenericMapHandler(string kind, function<Mapper(unordered_map<string, Scalar>)> handler)
```

**Pros:**
 * No need to write proto file
 * No need to create an exec node
**Cons:**
 * Only handles 1-in relations that don't accumulate

Reply:
Also, sorry, I was starting a bit with a blank slate above.  I think, in this PR, the current idea is that instead of using `GenericOptions` we use `string name` and encode the options into the name.  This is probably less flexible but in many cases users might be more familiar with encoding options into strings (ala URL encoding) instead of dealing with Arrow scalars.  I'm fine with either approach.

Reply:
So sticking with `name` I think the two approaches are:

### Generic Node
```
void RegisterGenericExtensionHandler(string kind, function<Declaration(string)> handler);
```
### Generic Mapper
```
using Mapper = function<ExecBatch(ExecBatch)>;
void RegisterGenericMapHandler(string kind, function<Mapper(string)> handler);
```

Reply:
Somehow missed this message - I am leaning towards "Generic Node" since it allows more flexibility to define node that is not just "map-like" and also we don't need to write protobuf file. 

I have a question - if we go the "Generic Node" route it looks like we can achieve register custom handlers via either `RegisterGenericExtensionHandler` or `set_default_extension_provider` [https://github.com/apache/arrow/commit/7423f0332cb11eb780f421c07bac71f87bf44a03#diff-8c5c127e8db8e52b40519262ee2b7e0179d4a39e5bcfe8b2165173b2b352d6a9R139](url) so technically we don't need `set_default_extension_provider` if we have `RegisterGenericExtensionHandler`?